shopinvader_api_payment/routers/payment.py

```diff
@@ -0,0 +1,188 @@
+
+import logging
+from typing import Annotated
+from urllib.parse import urljoin
+
+from fastapi import APIRouter, Depends, HTTPException, Request
+
+from odoo import api, models
+
+from odoo.addons.fastapi.dependencies import odoo_env
+from odoo.addons.payment.models.payment_provider import PaymentProvider
+
+from ..schemas import (
+    PaymentDataWithMethods,
+    PaymentProvider as PaymentProviderSchema,
+    TransactionCreate,
+    TransactionProcessingValues,
+)
+from .utils import Payable
+
+_logger = logging.getLogger(__name__)
+
+payment_router = APIRouter(tags=["payment"])
+
+
+@payment_router.get("/payment/methods")
+def pay(
+    payable: str,
+    odoo_env: Annotated[api.Environment, Depends(odoo_env)],
+) -> PaymentDataWithMethods:
+    """Available payment providers for the given encoded payment data.
+
+    This route is public, so it is possible to pay anonymously provided that the
+    parameters are obtained securely by another mean. An authenticated user can
+    obtain the parameters with corresponding routes on the related payable
+    objects (/cart/current/payable for e.g.).
+    """
+    try:
+        payable_obj = Payable.decode(odoo_env, payable)
+    except Exception as e:
+        _logger.info("Could not decode payable")
+        raise HTTPException(403) from e
+    # This method is similar to Odoo's PaymentPortal.payment_pay
+    providers_sudo = (
+        odoo_env["payment.provider"]
+        .sudo()
+        ._get_compatible_providers(
+            payable_obj.company_id,
+            payable_obj.partner_id,
+            payable_obj.amount,
+            currency_id=payable_obj.currency_id,
+        )
+    )
+    return PaymentDataWithMethods(
+        payable=payable,
+        payable_reference=payable_obj.payable_reference,
+        amount=payable_obj.amount,
+        currency_code=odoo_env["res.currency"].browse(payable_obj.currency_id).name,
+        amount_formatted=odoo_env["sale.order"]
+        .sudo()
+        .browse(payable_obj.payable_id)
+        .currency_id.format(payable_obj.amount),
+        providers=[
+            PaymentProviderSchema.from_payment_provider(provider)
+            for provider in providers_sudo
+        ],
+    )
+
+
+@payment_router.post("/payment/transactions")
+def transaction(
+    data: TransactionCreate,
+    request: Request,
+    odoo_env: Annotated[api.Environment, Depends(odoo_env)],
+) -> TransactionProcessingValues:
+    """Create a payment transaction.
+
+    Input is data obtained from /payment/providers, with the provider selected by the
+    user. This route is public, so it is possible to pay anonymously.
+
+    This route will automatically redirect to the return route linked to
+    the specified provider. The user will finally land on data.frontend_redirect_url
+    """
+    try:
+        payable_obj = Payable.decode(odoo_env, data.payable)
+    except Exception as e:
+        _logger.info("Could not decode payable")
+        raise HTTPException(403) from e
```

Comment:
What is the motivation for 403 and not 500 ?

Reply:
Because it is more likely an error in client input (4xx) than a server error (5xx).
And if the client attempts to send an invalid payable, it could be an attack, so we chose 403 forbidden.